draft-pda-core.md

```diff
@@ -166,24 +199,44 @@ limited set of aggregation functions, some of which we highlight below:
 This variety of aggregate types is sufficient to support a wide variety of
 data aggregation tasks.
 
-
-## Validating inputs
-
-A crucial task of any data collection pipeline is ensuring that the input data
-is valid. Going back to the example above, it's often useful to assert that each
-measurement is in a certain range, e.g., [0, 2^k) for some k. This straight-forward
-task is complicated in Prio by the fact that the inputs are secret shared. In
-particular, a malicious client can corrupt the computation by submitting random
-integers instead of a proper secret sharing of a valid input.
-
-To solve this problem, Prio introduces a light-weight zero-knowledge proof
-system designed to operate on secret shared data. In addition to its input
-share, each client sends to each aggregator a share of a "proof" of the input's
-validity. The aggregators use these proof shares in a protocol designed to
-establish the input's validity, without leaking the input itself. We describe
-this input-validation protocol in detail in [[TODO:citeme]].
-
-## Assembling Reports
+**Hits.**
+A common PA task that can't be solved efficiently with Prio is the
+`t`-*heavy-hitters* problem {{HITS}}. In this setting, each user is in
+possession of an `n`-bit string, and the goal is to compute the compute the set
+of strings that occur at least `t` times.
+
+[TODO: Provide an overview of the protocol of [BBG+21] and provide some
+intuition about how additive secret sharing is used. Be sure to introduce the
+notion of distributed point functions.]
+
+## Validating inputs in zero knowledge
+
+An essential task of any data collection pipeline is ensuring that the input
+data is "valid". Going back to the example above, it's often useful to assert
+that each measurement is in a certain range, e.g., `[0, 2^k)` for some `k`.
+This straight-forward task is complicated in our setting by the fact that the
+inputs are secret shared. In particular, a malicious client can corrupt the
+computation by submitting random integers instead of a proper secret sharing of
+a valid input.
+
+To solve this problem, each PA protocol in this document specifies a
+*zero-knowledge proof (ZKP) system* that allows the aggregators to verify that
+their shares correspond to as valid input. The system needs to have the
+following security properties (stated informally here; see {{BBG+19}} for a
+formal definition):
+1. *Completeness:* The verification procedure always succeeds on valid inputs.
```

Comment:
Can we move these to security considerations? I don't think they're important here.

Reply:
I disagree.

Reply:
FWIW, I agree with Wood.  The point here is that there is a proof system that demonstrates validity. The details should go below.

Reply:
I would push back as follows: several weeks ago I had to explain to @ekr that input validation fails to detect an invalid input with non-zero probability, but that this probability could be bounded. In general, this is a property of ZKP systems: the ZKP system is *sound* if verification correctly detects invalid inputs except with small probability. I don't think this is overly technical. Moreover, it's necessary to explain to the reader that input validation may fail.

Reply:
Nobody is saying that you don't say it. Rather, we're saying that you should say it *later*, because it's a detail and is not needed to understand the big picture.

The more detail you have in this section the harder it is for the reader to get the overall idea.

Reply:
In general, I think there's a danger here of contorting the language so much in order to avoid using "advanced" crypto terminology that we confuse the reader.

Reply:
That may be true, but I think Ekr's point is that we want to try and decrease cognitive load on the reader (implementer). They should exactly what they need to know to implement the protocol. Everything else can be described elsewhere!

Reply:
I think implementers are going to have to understand the term "ZKP" to the same depth they understand "encrypt" or "secret share". I'm defining "ZKP" here because w'er also defining "secret sharing" here. I don't think it makes sense to define one but not the other. They are both equally important at this stage.

In any case, this is a purely editorial point. I would suggest one of two paths forward: (1) we punt on this conversation and leave an "OPEN ISSUE" for cleaning up the presentation. (We'll have to do this at some point anyway.) (2) remove the text completely and leave a "TODO" for describing input validation.

Reply:
Or (3) does someone want to propose alternate text? The things I think are important to convey are (a) that the verification protocol does not leak the secret inputs to the aggregators and (b) the verification protocol may fail to detect invalid inputs with small probability.

Reply:
@cjpatton I think we're talking past each other: I agree that this is useful text, I just don't think it belongs *here*.

The generic structure of this kind of specification is:

- Problem statement
- Overview of the solution that just is enough to give people the idea
- Detailed protocol specification
- Security considerations and the like

The key idea with point (2) is to be high level. That means avoiding detail which isn't absolutely necessary. In particular, why do we need to tell users at this point that it might fail to detect invalid inputs? They don't need to know that to understand the protocol specification. See https://tools.ietf.org/rfcmarkup?doc=4101 for more on this.

Reply:
> The key idea with point (2) is to be high level. That means avoiding detail which isn't absolutely necessary. In particular, why do we need to tell users at this point that it might fail to detect invalid inputs?

I guess I don't quite understand what "absolutely necessary" means. It's important to know that the input validation protocol can be attacked, since mitigating such attacks is one of the core design criteria for this class of protocols. IMO we have to mention it up front.

Towards finding common ground, I'm wondering if it's strictly about using the term "ZKP"? Wood suggested that I might just be trying to use a fancy crypto term here. I assure you that I'm not. I'm just trying to provide the context one needs to understand the system. Using the term itself is useful because it connects what we're doing to a fairly well-known body of work. I don't want to give people the impression that the crypto herein is one-off.

> They don't need to know that to understand the protocol specification. See https://tools.ietf.org/rfcmarkup?doc=4101 for more on this.

So, up until this point, I haven't been thinking of this document as a formal standard. Based on the feedback I'm getting from you and Wood, it seems like I might need to recalibrate. I thought the goal of this change was merely to get us all on the same page about overall shape of the protocol, not (yet!) to provide a reference for others to implement. It's probably worth working this out in the next design call.

By the way, I would love to participate in the task of shaping this into a formal standard. I just didn't think that's what we were doing yet.

Reply:
I don't care much one way or the other about the term "zero-knowledge proof". I think you could say "proof" just as well. What I am saying is that your list of properties is unnecessary here. You can just say:

"To solve this problem, in each PA protocol, the client generates a
*zero-knowledge proof (ZKP)* of its input's validity, which the aggregators use
to verify that their shares correspond to as valid input."

And stop.  Or maybe say "This proof can be verified without learning the input".

You then move the properties to the Security Considerations. If you must, you can put a pointer to that here. Anyone who knows enough about ZKPs to care about the details can check there, and anyone who doesn't will read the text I have proposed and get substantively the right impression.


> Using the term itself is useful because it connects what we're doing to a fairly well-known body of work. I don't want to give people the impression that the crypto herein is one-off.

That can be done later in the specification

> So, up until this point, I haven't been thinking of this document as a formal standard. Based on the feedback I'm getting from you and Wood, it seems like I might need to recalibrate. I thought the goal of this change was merely to get us all on the same page about overall shape of the protocol, not (yet!) to provide a reference for others to implement. It's probably worth working this out in the next design call.

Well, I don't understand the point of this text at all if it's not targeted at an eventual specification. But yes, I do think this should be an attempt to write down pieces of a specification to use as input into the standards process.

Reply:
How's this:
```
To solve this problem, in each PA protocol, the client generates a
zero-knowledge proof of its input's validity that the aggregators use
to verify that their shares correspond to as valid input. The verification
procedure is designed to ensure that the aggregators learn nothing about the
input beyond its validity.
```
To me it's useful to use the term "zero-knowledge" because it signals what's under the hood. If you want I can remove it.

> Well, I don't understand the point of this text at all if it's not targeted at an eventual specification. But yes, I do think this should be an attempt to write down pieces of a specification to use as input into the standards process.

Of course that's what we're doing :) I meant more that I don't see why we need to settle on the presentation right now. I'd prefer to focus on the protocol itself and less on how it's presented. I figure that we will need to rework this doc quite a lot before we consider trying to get buy in from the wider community.